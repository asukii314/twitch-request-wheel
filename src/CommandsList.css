#commands-list,
#commands-list .nav-tabs,
#commands-list .nav-link {
    --bs-body-bg: #212529;
    --bs-link-color: #63a8fe;
    --bs-link-hover-color: #8bb9fe;
    --bs-secondary-color: rgba(222, 226, 230, 0.75);
    --bs-border-color: #495057;
    --bs-secondary-bg: #343a40;
    --bs-emphasis-color: #fff;
    --bs-nav-tabs-link-active-bg: #2f4f4f;
    --bs-nav-tabs-link-active-color: #fff;
    --bs-nav-tabs-border-color: #2f4f4f;
    --bs-nav-tabs-link-active-border-color: var(--bs-border-color) var(--bs-border-color) var(--bs-body-bg);
    --bs-nav-tabs-link-hover-border-color: var(--bs-border-color) var(--bs-border-color) var(--bs-body-bg);

}

.commands-table-wrapper {
    background-color: #2f4f4f;
}

#commands-list .text-primary-emphasis {
    color: #7fffd4!important;
}

#commands-list a.nav-link {
    border-bottom-color: var(--bs-nav-tabs-link-active-bg)!important;
}