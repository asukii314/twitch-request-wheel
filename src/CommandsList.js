import { Component } from 'react';
import YAML from 'yaml'
import rawCommandsList from './Commands.yaml';
import './CommandsList.css';
import {Nav, Table} from 'react-bootstrap';
const fetch = require('node-fetch');


export const SectionTable = function (data, i) {
    // console.log('SectionTable:', data);
    let {entries} = data;
    return (
        <Table striped bordered hover variant="dark" responsive="lg" key={`command-table-row-${i}`} className="mb-0">
            {/* <caption>{section}</caption> */}
            <thead>
                <tr className="text-center align-middle">
                    <th width="10%">Command</th>
                    <th width="10%">Permissions</th>
                    <th width="40%">Description</th>
                    <th width="20%">On Screen Equivalent</th>
                    <th width="10%">Variants</th>
                    <th width="10%">Example</th>
                </tr>
            </thead>
            <tbody className="table-group-divider">
                {!!entries && entries.map(CommandTableRow)}
            </tbody>
        </Table>

    );
}
export const CommandTableRow = function ({command, info}, i) {
    const {
        Example,
        Availability,
        Description,
        OnScreenEquivalent,
        Variants,
    } = info;

    return (
        <tr key={`command-table-row-${i}`} className="small">
            <td className="command-name fw-semibold fs-6 text-primary-emphasis">!{command}</td>
            <td className="text-center">{Availability}</td>
            <td>{Description}</td>
            <td>{OnScreenEquivalent}</td>
            <td className={Variants==='n/a' ? null : 'font-monospace fst-italic'}>{Variants}</td>
            <td className="font-monospace fst-italic">{Example}</td>
        </tr>
    );
}

export default class CommandsList extends Component {
    constructor(props) {
        super(props);
        this.state = {
            activeFilter: null,
            validCommands: []
        };
    }

    componentDidMount(props) {
        fetch(rawCommandsList)
            .then(r => r.text())
            .then(text => {
                let validCommands = YAML.parse(text);
                this.setState((state) => {
                    return {
                        ...state,
                        validCommands: Object.entries(validCommands).map(
                            ([section, commands], i) => {
                                return {
                                    section,
                                    commands,
                                    entries: Object.entries(commands).map(([command, info], i) => ({command, info})),
                                };
                            }
                        )
                    };
                });
            });
    }

    onSelectTab = (eventKey) => {
        switch (eventKey) {
            case 'All Commands':
                return this.setState({
                    activeFilter: null
                });
            default:
                return this.setState({
                    activeFilter: eventKey
                });
        }
    }

    render() {
        const {validCommands} = this.state;

        let navItems = [
            {section: 'All Commands'},
            ...validCommands
        ].map(cmd => {
            const {section} = cmd;
            return (
                <Nav.Item key={section}>
                    <Nav.Link eventKey={section} data-testid={section}>
                        {section}
                    </Nav.Link>
                </Nav.Item>
            );
        });

        return (
            <div id="commands-list" className="container">
                <h1 className="fw-bolder pt-3">Chat Commands</h1>
                <Nav variant="tabs" activeKey={this.state.activeFilter || 'All Commands'} onSelect={this.onSelectTab}>
                    {/* <Nav.Item>
                        <Nav.Link eventKey="All Commands" data-testid="All Commands">All Commands</Nav.Link>
                    </Nav.Item> */}
                    {validCommands.length > 0 && navItems}
                </Nav>
                <div className="col-12 px-3 pb-1 mb-3 text-start commands-table-wrapper">
                    {!!validCommands && validCommands.map(
                        (data, i) => {
                            let {activeFilter} = this.state;
                            if (activeFilter && activeFilter !== data.section) {
                                return null;
                            }
                            return (
                                <div key={`section-${i}`}>
                                    <h4 className="px-1 pt-3 fw-bolder" key={`section-table-row-${i}`}>
                                        {data.section}
                                    </h4>
                                    {SectionTable(data)}
                                </div>
                            );
                        }
                    ).filter(cmd => cmd)}
                </div>
            </div>
        );
    }
}