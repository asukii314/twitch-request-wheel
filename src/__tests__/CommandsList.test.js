import CommandsList from '../CommandsList';
import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import fetch from 'node-fetch';
jest.mock('node-fetch');

const yamlCommandsList = `
Game Requests:
  request:
    Example: '!request MSM'
    Availability: 'Everyone'
    Description: 'Adds the requested game to the wheel. (If you already had a different active game request, it will be replaced - unless you are the broadcaster.)'
    OnScreenEquivalent: 'n/a'
    Variants: '!request'

  subrequest:
    Example: '!subrequest MSM'
    Availability: 'Subscribers'
    Description: 'Adds the requested game to the wheel in a separate section when enabled in settings; this will not replace a previously request game using the !request command. (If you already had a different active game request, it will be not replaced unless Limit 1 Sub Request is enabled in streamer settings.)'
    OnScreenEquivalent: 'n/a'
    Variants: 'n/a'

Seat Requests:
  new:
    Example: '!new'
    Availability: 'Everyone'
    Description: 'Adds the requesting user to the "interested" column. (If the !caniplay command is used, a confirmation message will be sent in chat.)'
    OnScreenEquivalent: 'n/a'
    Variants: '!caniplay'

Other / Misc:
  whichpack:
    Example: '!whichpack TMP 2'
    Availability: 'Everyone'
    Description: 'Looks up the specified game and responds with which party pack (if any) it is in.'
    OnScreenEquivalent: 'n/a'
    Variants: 'n/a'
`;

describe('CommandsList', () => {
    describe('onSelectTab', () => {
        test('should handle all cases', () => {
            let component = new CommandsList({});
            jest.spyOn(component, 'setState').mockImplementation(()=>{});
            component.onSelectTab('All Commands');
            component.onSelectTab('Game Commands');
            // console.log('component.setState.mock.calls:', component.setState.mock.calls);
            expect(component.setState.mock.calls[0]).toEqual([{
                activeFilter: null
            }], [{
                activeFilter: 'Game Commands'
            }])
        });
    });
    describe('render', () => {
        let props;
        beforeEach(()=>{
            fetch.mockImplementation(() => {
                return Promise.resolve({
                    text: () => Promise.resolve(yamlCommandsList),
                })
            });

            props = {};
        });

        test('renders component with list of all commands using yaml content', async () => {
            let {container} = render(<CommandsList {...props} />);

            await waitFor(() => screen.getByText(/All Commands/i));
            expect(screen.queryByText(/!subrequest MSM/i)).toBeInTheDocument();
            expect(screen.queryByText(/!whichpack TMP 2/i)).toBeInTheDocument();
            expect(container).toMatchSnapshot();
        });
        test('renders component with filtered list of commands yaml content', async () => {
            render(<CommandsList {...props} />);

            await waitFor(() => screen.getByText(/All Commands/i));

            fireEvent.click(screen.getByTestId('Other / Misc'));

            await waitFor(() => screen.getByText(/All Commands/i));
            expect(screen.queryByText(/!subrequest MSM/i)).not.toBeInTheDocument();
            expect(screen.queryByText(/!whichpack TMP 2/i)).toBeInTheDocument();

            fireEvent.click(screen.getByText(/All Commands/i));

            await waitFor(() => screen.getByText(/All Commands/i));
            expect(screen.queryByText(/!subrequest MSM/i)).toBeInTheDocument();
            expect(screen.queryByText(/!whichpack TMP 2/i)).toBeInTheDocument();
        });
    });

});
