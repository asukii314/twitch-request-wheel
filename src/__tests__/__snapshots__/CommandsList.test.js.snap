// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandsList render renders component with list of all commands using yaml content 1`] = `
<div>
  <div
    class="container"
    id="commands-list"
  >
    <h1
      class="fw-bolder pt-3"
    >
      Chat Commands
    </h1>
    <div
      class="nav nav-tabs"
    >
      <div
        class="nav-item"
      >
        <a
          class="nav-link active"
          data-rr-ui-event-key="All Commands"
          data-testid="All Commands"
          href="#"
          role="button"
          tabindex="0"
        >
          All Commands
        </a>
      </div>
      <div
        class="nav-item"
      >
        <a
          class="nav-link"
          data-rr-ui-event-key="Game Requests"
          data-testid="Game Requests"
          href="#"
          role="button"
          tabindex="0"
        >
          Game Requests
        </a>
      </div>
      <div
        class="nav-item"
      >
        <a
          class="nav-link"
          data-rr-ui-event-key="Seat Requests"
          data-testid="Seat Requests"
          href="#"
          role="button"
          tabindex="0"
        >
          Seat Requests
        </a>
      </div>
      <div
        class="nav-item"
      >
        <a
          class="nav-link"
          data-rr-ui-event-key="Other / Misc"
          data-testid="Other / Misc"
          href="#"
          role="button"
          tabindex="0"
        >
          Other / Misc
        </a>
      </div>
    </div>
    <div
      class="col-12 px-3 pb-1 mb-3 text-start commands-table-wrapper"
    >
      <div>
        <h4
          class="px-1 pt-3 fw-bolder"
        >
          Game Requests
        </h4>
        <div
          class="table-responsive-lg"
        >
          <table
            class="mb-0 table table-dark table-striped table-bordered table-hover"
          >
            <thead>
              <tr
                class="text-center align-middle"
              >
                <th
                  width="10%"
                >
                  Command
                </th>
                <th
                  width="10%"
                >
                  Permissions
                </th>
                <th
                  width="40%"
                >
                  Description
                </th>
                <th
                  width="20%"
                >
                  On Screen Equivalent
                </th>
                <th
                  width="10%"
                >
                  Variants
                </th>
                <th
                  width="10%"
                >
                  Example
                </th>
              </tr>
            </thead>
            <tbody
              class="table-group-divider"
            >
              <tr
                class="small"
              >
                <td
                  class="command-name fw-semibold fs-6 text-primary-emphasis"
                >
                  !
                  request
                </td>
                <td
                  class="text-center"
                >
                  Everyone
                </td>
                <td>
                  Adds the requested game to the wheel. (If you already had a different active game request, it will be replaced - unless you are the broadcaster.)
                </td>
                <td>
                  n/a
                </td>
                <td
                  class="font-monospace fst-italic"
                >
                  !request
                </td>
                <td
                  class="font-monospace fst-italic"
                >
                  !request MSM
                </td>
              </tr>
              <tr
                class="small"
              >
                <td
                  class="command-name fw-semibold fs-6 text-primary-emphasis"
                >
                  !
                  subrequest
                </td>
                <td
                  class="text-center"
                >
                  Subscribers
                </td>
                <td>
                  Adds the requested game to the wheel in a separate section when enabled in settings; this will not replace a previously request game using the !request command. (If you already had a different active game request, it will be not replaced unless Limit 1 Sub Request is enabled in streamer settings.)
                </td>
                <td>
                  n/a
                </td>
                <td>
                  n/a
                </td>
                <td
                  class="font-monospace fst-italic"
                >
                  !subrequest MSM
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div>
        <h4
          class="px-1 pt-3 fw-bolder"
        >
          Seat Requests
        </h4>
        <div
          class="table-responsive-lg"
        >
          <table
            class="mb-0 table table-dark table-striped table-bordered table-hover"
          >
            <thead>
              <tr
                class="text-center align-middle"
              >
                <th
                  width="10%"
                >
                  Command
                </th>
                <th
                  width="10%"
                >
                  Permissions
                </th>
                <th
                  width="40%"
                >
                  Description
                </th>
                <th
                  width="20%"
                >
                  On Screen Equivalent
                </th>
                <th
                  width="10%"
                >
                  Variants
                </th>
                <th
                  width="10%"
                >
                  Example
                </th>
              </tr>
            </thead>
            <tbody
              class="table-group-divider"
            >
              <tr
                class="small"
              >
                <td
                  class="command-name fw-semibold fs-6 text-primary-emphasis"
                >
                  !
                  new
                </td>
                <td
                  class="text-center"
                >
                  Everyone
                </td>
                <td>
                  Adds the requesting user to the "interested" column. (If the !caniplay command is used, a confirmation message will be sent in chat.)
                </td>
                <td>
                  n/a
                </td>
                <td
                  class="font-monospace fst-italic"
                >
                  !caniplay
                </td>
                <td
                  class="font-monospace fst-italic"
                >
                  !new
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div>
        <h4
          class="px-1 pt-3 fw-bolder"
        >
          Other / Misc
        </h4>
        <div
          class="table-responsive-lg"
        >
          <table
            class="mb-0 table table-dark table-striped table-bordered table-hover"
          >
            <thead>
              <tr
                class="text-center align-middle"
              >
                <th
                  width="10%"
                >
                  Command
                </th>
                <th
                  width="10%"
                >
                  Permissions
                </th>
                <th
                  width="40%"
                >
                  Description
                </th>
                <th
                  width="20%"
                >
                  On Screen Equivalent
                </th>
                <th
                  width="10%"
                >
                  Variants
                </th>
                <th
                  width="10%"
                >
                  Example
                </th>
              </tr>
            </thead>
            <tbody
              class="table-group-divider"
            >
              <tr
                class="small"
              >
                <td
                  class="command-name fw-semibold fs-6 text-primary-emphasis"
                >
                  !
                  whichpack
                </td>
                <td
                  class="text-center"
                >
                  Everyone
                </td>
                <td>
                  Looks up the specified game and responds with which party pack (if any) it is in.
                </td>
                <td>
                  n/a
                </td>
                <td>
                  n/a
                </td>
                <td
                  class="font-monospace fst-italic"
                >
                  !whichpack TMP 2
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
